### PR TITLE
[CFM_110] Update Fridge  with latest report when there's a new fridge report

### PR DESCRIPTION
**Pull request by 1jeanpaul1**
Trello Ticket: https://trello.com/c/wObHVifP/118-cfm110-update-fridge-table-with-latest-report-when-a-new-fridgereport-is-made
### Updates
1. [CommunityFridgeMapApi/dependencies/python/db.py](https://github.com/CollectiveFocus/CFM_Backend/compare/CFM_110_JP?expand=1#diff-9c2aafe3c6ce0facf2c86db6635c7f7c95cd4eeddeceaab949b2da0033f7f44e)
    * Added `update_fridge_report()` to Fridge class. Updates latestFridgeReport field and json_data
    * Updated FridgeReport class to work with `format_dynamodb_item_v2()`. Added `FIELD_VALIDATION` field
    * Renamed FridgeReport fields to be compliant with our api contract
2. Updated unit tests
3. Renamed **fridge_id** to **fridgeId** to be consistent throughout the project

### Unit Tests
1. Run Unit Tests: `coverage run -m pytest tests/unit -v`

### QA

* Pre-Requisite
   * build container: `sam build --use-container`
   * Run dynamodb locally, create tables, and populate fridge table. Follow README instructions to do so if you haven't already https://github.com/CollectiveFocus/CFM_Backend/blob/dev/README.md#setup-local-database-connection
   

#### One Time Use
1. POST FridgeReport: `sam local invoke FridgeReportFunction --event events/local-fridge-report-event.json --parameter-overrides ParameterKey=Environment,ParameterValue=local --docker-network cfm-network`
    * [OPTIONAL] Generate custom event Example: `sam local generate-event apigateway aws-proxy --method POST --path document --body "{\"status\": \"working\", \"fridge_percentage\": 0}" > events/local-fridge-report-event-2.json`
        * Add `"fridgeId": "{FRIDGEID}"` to pathParameter in generated file
2. Query Data: `aws dynamodb scan --table-name fridge_report --endpoint-url http://localhost:8000`

#### Local Server
1. Start Server: `sam local start-api --parameter-overrides ParameterKey=Environment,ParameterValue=local --docker-network cfm-network`
2. Make a POST Request to: `http://127.0.0.1:3000/v1/fridges/{fridgeId}/reports`
    * Example: `curl --location --request POST 'http://127.0.0.1:3000/v1/fridges/thefriendlyfridge/reports' --header 'Content-Type: application/json' --data-raw '{"status": "working", "fridge_percentage": 100}'`
4. Confirm that `thefriendlyfridge` fridge `latestFridgeReport` and `json_data` field were updated

#### Confirmation

1. Run Local Server: sam local start-api --parameter-overrides ParameterKey=Environment,ParameterValue=local --docker-network cfm-network
2. Confirm that `2fish5loavesfridge` fridge `latestFridgeReport` and `json_data` field were updated
    * Go to http://localhost:3000/v1/fridges/2fish5loavesfridge
3. Confirm that  `thefriendlyfridge` fridge `latestFridgeReport` and `json_data` field were updated
    * Go to http://localhost:3000/v1/fridges/thefriendlyfridge

### CODE REVIEW

File: CommunityFridgeMapApi/tests/unit/test_fridge_report.py [hunk 1]
```diff
@@ -9,130 +9,171 @@ def __init__(self):
...
-        self.assertEqual(field, "fridge_percentage")
+    def update_item(
+        self,
+        TableName=None,
+        Key=None,
+        ExpressionAttributeNames=None,
+        ExpressionAttributeValues=None,
+        UpdateExpression=None,
+    ):
+        pass
```
Comment (abhinav1912) [inline CommunityFridgeMapApi/tests/unit/test_fridge_report.py:23]: Why isn't there a test for this method?

File: CommunityFridgeMapApi/dependencies/python/db.py [hunk 2]
```diff
@@ -471,35 +472,84 @@ def add_item(self, conditional_expression=None) -> DB_Response:
...
+        latestFridgeReport = json.dumps(fridge_report)
+        self.db_client.update_item(
+            TableName=self.TABLE_NAME,
+            Key={"id": {"S": fridgeId}},
+            ExpressionAttributeNames={"#LFR": "latestFridgeReport", "#JD": "json_data"},
+            ExpressionAttributeValues={
+                ":fr": {"S": latestFridgeReport},
+                ":fj": {"S": fridge_json_data},
+            },
+            UpdateExpression="SET #JD = :fj, #LFR = :fr",
```
Comment (abhinav1912) [inline CommunityFridgeMapApi/dependencies/python/db.py:496]: nit: can you explain this expression?

**Comment by abhinav1912** [comment]
> Renamed fridge_id to fridgeId to be consistent throughout the project

Why though? I understand this change in `.yml` or `JSON` objects, but snake case is the more _pythonic_ way. (Not a big deal, just asking).

File: CommunityFridgeMapApi/tests/unit/test_fridge_report.py (hunk 1 above)
Comment (CollectiveFocus) [reply CommunityFridgeMapApi/tests/unit/test_fridge_report.py:23]: The diff makes this look weird. This is a function for mock `DynamoDbMockPutItem` so it doesn't need a test

**Comment by CollectiveFocus** [comment]
> > Renamed fridge_id to fridgeId to be consistent throughout the project
> 
> Why though? I understand this change in `.yml` or `JSON` objects, but snake case is the more _pythonic_ way. (Not a big deal, just asking).

I agree, and I would prefer **fridge_id** but I kept getting errors for writing the wrong field so to avoid the headache I made them all the same 


Also the way we implemented **format_dynamodb_item_v2** made it so the constructor had to be in the form `fridgeId` to be compliant with the API contract https://github.com/CollectiveFocus/CFM_Backend/pull/19/files#diff-9c2aafe3c6ce0facf2c86db6635c7f7c95cd4eeddeceaab949b2da0033f7f44eR194. So since that was a requirement, I sided with making them all the same format

File: CommunityFridgeMapApi/dependencies/python/db.py (hunk 2 above)
Comment (CollectiveFocus) [reply CommunityFridgeMapApi/dependencies/python/db.py:496]: Yeah they make it really confusing how you have to do this. The goal is to get here:
`SET json_data = {"S": fridge_json_data}, "#LFR": {"S": latestFridgeReport}`

but to get there we need to write something for **ExpressionAttributeNames** and **ExpressionAttributeValues**

------
Here's my best attempt at explaining it, lmk if its good and I'll add it in as a comment:
```
# JD and #LFR are mapped to the values set in ExpressionAttributeNames
# :fj and :fr are mapped to the values set in ExpressionAttributeValues
UpdateExpression becomes: "json_data":  {"S": fridge_json_data}, "latestFridgeReport": {"S": latestFridgeReport}
```

Comment (CollectiveFocus) [reply CommunityFridgeMapApi/dependencies/python/db.py:501]: done: 03ae4a50c0605ce95deb8658d6baf5a9a4183e16

**Review by abhinav1912 — APPROVED** [review]

**Review by CollectiveFocus — APPROVED** [review]

**Pull request merged**